csharp/src/Apache.Arrow.Adbc/Tracing/ActivityWithPii.cs

```diff
@@ -0,0 +1,225 @@
+            _activity?.AddLink(new ActivityLink(ActivityContext.Parse(traceParent, null), tags: tagsCollection));
+            return this;
+        }
+
+        /// <summary>
+        /// Update the Activity to have a tag with an additional 'key' and value 'value'.
+        /// This shows up in the <see cref="TagObjects"/> enumeration. It is meant for information that
+        /// is useful to log but not needed for runtime control (for the latter, <see cref="Baggage"/>)
+        /// </summary>
+        /// <returns><see langword="this" /> for convenient chaining.</returns>
+        /// <param name="key">The tag key name</param>
+        /// <param name="value">The tag value mapped to the input key</param>
+        /// <param name="isPii">Indicates whether the value is PII (Personally Identifiable Information).</param>
+        public ActivityWithPii AddTag(string key, object? value, bool isPii = true)
+        {
+            value = MaybeToRedactedValue(value, isPii);
+            _activity.AddTag(key, value);
+            return this;
+        }
+
+        /// <summary>
+        /// Update the Activity to have a tag with an additional 'key' and value 'value'.
+        /// This shows up in the <see cref="TagObjects"/> enumeration. It is meant for information that
+        /// is useful to log but not needed for runtime control (for the latter, <see cref="Baggage"/>)
+        /// </summary>
+        /// <returns><see langword="this" /> for convenient chaining.</returns>
+        /// <param name="key">The tag key name</param>
+        /// <param name="valueFn">The tag value function mapped to the input key</param>
+        /// <param name="isPii">Indicates whether the value is PII (Personally Identifiable Information).</param>
+        public ActivityWithPii AddTag(string key, Func<object?> valueFn, bool isPii = true)
+        {
+            object? value = MaybeToRedactedValue(valueFn(), isPii);
+            _activity.AddTag(key, value);
+            return this;
+        }
+
+        /// <summary>
+        /// Update the Activity to have a tag with an additional 'key' and value 'value'.
+        /// This shows up in the <see cref="TagObjects"/> enumeration. It is meant for information that
+        /// is useful to log but not needed for runtime control (for the latter, <see cref="Baggage"/>)
+        /// </summary>
+        /// <returns><see langword="this" /> for convenient chaining.</returns>
+        /// <param name="key">The tag key name</param>
+        /// <param name="value">The tag value mapped to the input key as a function</param>
+        /// <param name="guidFormat">The format indicator for 16-byte GUID arrays.</param>
+        public ActivityWithPii AddTag(string key, byte[]? value, string? guidFormat, bool isPii = true)
+        {
+            if (value == null)
+            {
+                return AddTag(key, value, isPii);
+            }
+            if (value.Length == 16)
```

Comment:
This should check if guidFormat is non-null? It is strange that having a byte array of length 16 is automatically a GUID.